Adding new features to version 1.1.0

jsccjj opened the pull request:
Hi Bart,

I spent some free time looking into your wonderful package. I modified some codes and added new features based on version 1.1.0.

Major modifications (in my opinion):

1. Upgraded Blockly to version "5.20210325.1"
2. Added an Ace XML editor. Now load-from/save-to library is enabled. Saved files are XML that can properly re-create Blockly workspace
3. Created an expand button for the Blockly workspace using Red.trays api

Minor modifications:

1. When loading XML toolbox files, parse the categories in a forward fashion instead of backward so that nested categories can be parsed properly
2. used Blockly.Xml.domToPrettyText instead of Blockly.Xml.domToText

 I hope this may add value to you package.

Thanks, 
Jeff

bartbutenaers commented:
Morning Jeff,

A fullscreen-mode was for me the most important requirement, but I was completely stuck with.  Even with the help from the Blockly team (see [issue 1](https://github.com/google/blockly/issues/4382) and [issue 2](https://github.com/google/blockly/issues/4383)).  As a result the devolpment of this node was completely halted, and I had to disappoint my partner 
 Simon (@cymplecy) several times.  I think he already gave up all hope for a new major version of this node...

But seems you have managed to implement this feature.  Awesome!!!  Thank you so much!!  Moreover this is the first large pull request that I have ever received on any of my nodes...

I am not developing at the moment, since I'm recovering from an operation.  But if everything goes well, I hope to start working on a new release of this node in around 6 weeks from now.

I have not reviewed your code in detail, but could you: 
+ Add yourself to the [contributor](https://github.com/bartbutenaers/node-red-contrib-blockly/blob/master/package.json#L10) section.  I think you have earned that title ...
+ Seems you have added some extra blocks.  Can you explain those, so Simon can give his 'functional' feedback.
+ You expand only the Blockly editor tabsheet.  I think it is usefull that the full-screen mode contains the 3 tabsheets.  Suppose I want to learn programming Javascript.  I start with this node, and every time that I have dragged a block in the editor, I want to switch to the "Javascript" tabsheet over and over again.  But this now means I have to leave and enter your fullscreen mode every time, which is not really user friendly.  Is it possible to show the tabsheet control with 3 tabsheets in full-screen??
+ You have upgraded the Blockly version.  Do you have seen any other new features that we should support?

I will try to review your changes in detail as soon as possible.  Will get back to you with my feedback...

Have a nice sunday,
Bart

jsccjj commented:
Hi Bart,

Thank you for your recognition! I am glad that I could contribute. Also, my best wishes to you for your recovery from the operation.

Please see my responses below:

_**- Add yourself to the contributor section. I think you have earned that title ...**_

Jeff: My pleasure!

_**- Seems you have added some extra blocks. Can you explain those, so Simon can give his 'functional' feedback.**_

Jeff: I removed most of the added blocks since they are only useful for my building automation projects.
I only kept "Javascript Expression" block<br>
![image](https://user-images.githubusercontent.com/79299554/116040566-08ff1b00-a621-11eb-88a9-0d9c428007d2.png)
This is a, I think, a complementary block that is useful while users want to use methods such as isNaN(), isFinite(), and etc with Blockly core blocks.

_**- You expand only the Blockly editor tabsheet. I think it is usefull that the full-screen mode contains the 3 tabsheets. Suppose I want to learn programming Javascript. I start with this node, and every time that I have dragged a block in the editor, I want to switch to the "Javascript" tabsheet over and over again. But this now means I have to leave and enter your fullscreen mode every time, which is not really user friendly. Is it possible to show the tabsheet control with 3 tabsheets in full-screen??**_

Jeff: I created the expand view by using Red.tray() directly. Creating tabs in the expanded view reduce the editor area and make the code much longer since we have to redefine tabs as well as html part of it. So, I just created an alternative by using buttons:<br>
![image](https://user-images.githubusercontent.com/79299554/116041474-17016b80-a622-11eb-8416-f147cb5f4721.png)<br>
By clicking "See Generated Javascript" button, users can get to the readonly Javascript editor:<br>
![image](https://user-images.githubusercontent.com/79299554/116041685-57f98000-a622-11eb-959a-5d8d58461bbb.png)<br>
Users can get back to the Blockly workspace by clicking "Back to Blockly Editor".

I have committed this change to my fork. I will create another pull request if you think it's appropriate. 

_**- You have upgraded the Blockly version. Do you have seen any other new features that we should support?**_

Jeff: I have not investigated into it yet. Is there any features you are expecting?

bartbutenaers commented:
>  So, I just created an alternative by using buttons

That looks very user-friendly to me.  Nice development work!!!

>  I will create another pull request if you think it's appropriate.

No please do it all in this PR.  That makes it easier for me to review it afterwards.  

> Is there any features you are expecting?

I was just wondering if you had upgraded a new version of Blockly because you required new features.  But now I assume you just wanted to get my node in sync with Blockly.

Last year I had already implemented some changes (in a local version on my pc):
+ Extra nodes since the API of the Function-node sandbox has been extended with new features.
+ Removed the Blockly libs from this repo, and implemented a [Blockly NPM](https://www.npmjs.com/package/blockly) dependency (so new minor Blockly versions are automatically installed).

I will commit those changes to Github when I'm back into business...

cymplecy commented:
Hi Jeff
1st thoughts
![image](https://user-images.githubusercontent.com/2357428/116090661-04c41380-a69c-11eb-9ffe-a16b28ddb0cf.png)

Is different order - could it be changed to the original order
![image](https://user-images.githubusercontent.com/2357428/116090892-3f2db080-a69c-11eb-9ae4-0312f5fc4545.png)

Also, the resizing buttons could do with being fully visible without need to scroll
![image](https://user-images.githubusercontent.com/2357428/116093182-5e2d4200-a69e-11eb-97e9-b6b9e63ab806.png)

cymplecy commented:
On a separate issue the extra JavaScript function block comes up from time to time but it's never made it into the project as it's not really needed.

If you think that you have a use case that requires it - let me know what it is and I'll show you that it isn't needed :)

Simon

cymplecy commented:
And your JS blocks in your example aren't needed :)

![image](https://user-images.githubusercontent.com/2357428/116094870-cf212980-a69f-11eb-928f-98138fbbf96f.png)

jsccjj commented:
> > So, I just created an alternative by using buttons
> 
> That looks very user-friendly to me. Nice development work!!!
> 
Thank you!
> > I will create another pull request if you think it's appropriate.
> 
> No please do it all in this PR. That makes it easier for me to review it afterwards.
> 
Got it. I just double checked that the pull request is reflecting my latest commits.
> > Is there any features you are expecting?
> 
> I was just wondering if you had upgraded a new version of Blockly because you required new features. But now I assume you just wanted to get my node in sync with Blockly.
> 
> Last year I had already implemented some changes (in a local version on my pc):
> 
> * Extra nodes since the API of the Function-node sandbox has been extended with new features.
> * Removed the Blockly libs from this repo, and implemented a [Blockly NPM](https://www.npmjs.com/package/blockly) dependency (so new minor Blockly versions are automatically installed).
> 
> I will commit those changes to Github when I'm back into business...
I look forward to it! Let us know what else we can contribute.

jsccjj commented:
> Hi Jeff
> 1st thoughts
> ![image](https://user-images.githubusercontent.com/2357428/116090661-04c41380-a69c-11eb-9ffe-a16b28ddb0cf.png)
> 
> Is different order - could it be changed to the original order
> ![image](https://user-images.githubusercontent.com/2357428/116090892-3f2db080-a69c-11eb-9ae4-0312f5fc4545.png)
> 
Hi Simon, I changed the order per your comment but left the "Misc" at the bottom for the reason that the toolbox of "Misc" contains a "Logic" type of block. When joining the toolboxes this may lead all "Logic" blocks into "Misc" category. Of course this can be changed again if you think it's more proper. New look is here:
![image](https://user-images.githubusercontent.com/79299554/116156401-448cfa00-a6a0-11eb-940f-bc129dbc9f11.png)
<br>




 

> Also, the resizing buttons could do with being fully visible without need to scroll
> ![image](https://user-images.githubusercontent.com/2357428/116093182-5e2d4200-a69e-11eb-97e9-b6b9e63ab806.png)<br>

This part is fixed by changing the CSS style. The min-height setting of the expanded blockly view was set to 700px. Now, it is 350px. This should fix the problem. 
Please let me know whether this works at your end.

jsccjj commented:
> On a separate issue the extra JavaScript function block comes up from time to time but it's never made it into the project as it's not really needed.
> 
> If you think that you have a use case that requires it - let me know what it is and I'll show you that it isn't needed :)
> 
> Simon

The "Javascript Statement" block is powerful already. 

The situation I have is like:

Considering to generate the following code by blockly

if ((msg['topic']) == 'disabled_op') {
    disabled_op = Number(msg.payload);
    if (isNaN(disabled_op)) {
      disabled_op = 0;
    }
}

We can simply use the Statement block like this:
![image](https://user-images.githubusercontent.com/79299554/116159068-76a05b00-a6a4-11eb-9589-e829c4311662.png)
and get this:
![image](https://user-images.githubusercontent.com/79299554/116159110-891a9480-a6a4-11eb-95fd-bad27c4b0608.png)

Below would get the same thing
![image](https://user-images.githubusercontent.com/79299554/116159262-c3843180-a6a4-11eb-9af9-032848ae13d0.png)
By using "Javascript Expression", I would need additional blocks to achieve the same goal. However, I view this is the purpose of Blockly that is to visually learn coding. If I had know how to write all the JS statements, I would just use the default function node. 

Also, in the situation below, the expression block may be useful:
![image](https://user-images.githubusercontent.com/79299554/116162348-6e4b1e80-a6aa-11eb-9085-a1e8698bb80f.png)
![image](https://user-images.githubusercontent.com/79299554/116162393-8ae75680-a6aa-11eb-9741-93b11f45df72.png)

My two pennies here...

Thanks,
Jeff

cymplecy commented:
"Please let me know whether this works at your end."
The layout changes look/work fine for me :)

jsccjj commented:
Hi Simon,
I understand. I will remove the node.

bartbutenaers commented:
Hi Jeff,

Simon is completely right that the purpose of this node is an alternative for the function node, for those that don't feel confident with Javascript.  But on the other hand we understand your use case.

So we had a very short discussion about it, and we agreed that you can keep your Expression-node.  Consider it as a thank-you gift from our side, for all the time you have spend to help us improving this node ...

Hopefully you still have the source code somewhere ;-)

jsccjj commented:
> So we had a very short discussion about it, and we agreed that you can keep your Expression-node. Consider it as a thank-you gift from our side, for all the time you have spend to help us improving this node ...
> 

Thank you Bart and Simon. I appreciate you guys' approval! I will add the code back :)

BTW, have you guys checked the "Multiline text input fields" which seems added to blockly last Nov.
https://developers.google.com/blockly/guides/create-custom-blocks/fields/built-in-fields/multiline-text-input

![image](https://user-images.githubusercontent.com/79299554/116511790-d6049380-a87b-11eb-8dc3-d66338a160b3.png)

It seems it can be used on Comment and Statement blocks but may be incompatible to current blocks...

bartbutenaers commented:
@cymplecy,
Is it ok for you if @jsccjj adds the multiline text input block in our toolbox?  
I don't have any objections...

cymplecy commented:
Fine by me :)

jsccjj commented:
Hi guys,

I added the new block. Please give it a try.

the default:
![image](https://user-images.githubusercontent.com/79299554/116844321-a87b5b00-ab97-11eb-8b81-b256ff464c2a.png)
the generated JS:
![image](https://user-images.githubusercontent.com/79299554/116844354-bd57ee80-ab97-11eb-82f6-1373aa6d4410.png)


In action:
![image](https://user-images.githubusercontent.com/79299554/116844442-f5f7c800-ab97-11eb-8065-2c6b0ab8a9a9.png)
The generated JS:
![image](https://user-images.githubusercontent.com/79299554/116844466-027c2080-ab98-11eb-955a-eb3e0a25506b.png)

Please let me know how you want to improve the block once you have tried it.

Thanks, Jeff

cymplecy commented:
Hi Jeff
I'm having an issue when using normal editor (not expanded one)
If make edits in Editor mode and then switch to Generated JavaScript mode and switch back to Editor mode - the edits are lost.
Simon

jsccjj commented:
Hi Simon, 
That was a bug. I got it fixed 2 days ago. Please clone my fork  (https://github.com/jsccjj/node-red-contrib-blockly). And let me know if it works for you.  
Sorry for the bug.

Thanks, Jeff

cymplecy commented:
I thought I was on the latest but I hadn't cleared my cache :)

All OK now :)

bartbutenaers commented on README.md:
```diff
@@ -1,3 +1,41 @@
+
+This is a derived version of node-red-contrib-blockly release-1.1.0 @ https://github.com/bartbutenaers/node-red-contrib-blockly
+This version is a progressing experiment with upgraded Blockly, expanded Blockly editor, import/export XML library and etc.
+
```
Can you please remove this section on the readme, since it is related to your fork.

jsccjj replied:
Done. :)

bartbutenaers commented on lib/basic/msg/js/en.js:
```diff
@@ -2,429 +2,433 @@
+Blockly.Msg["PROCEDURES_DEFRETURN_TITLE"] = Blockly.Msg["PROCEDURES_DEFNORETURN_TITLE"];
+Blockly.Msg["TEXT_APPEND_VARIABLE"] = Blockly.Msg["VARIABLES_DEFAULT_NAME"];
+Blockly.Msg["TEXT_CREATE_JOIN_ITEM_TITLE_ITEM"] = Blockly.Msg["VARIABLES_DEFAULT_NAME"];
 
```
What is the reason for adding all those messages?
Because all the messages in this file need to be translated by volunteers...

jsccjj replied:
This is caused by using the updated "en.js":
https://github.com/google/blockly/blob/master/msg/js/en.js

bartbutenaers replied:
Ah damn.  After 3 years neglecting this node, I was confused by my own en.js and this one from Google.  Fine!

bartbutenaers commented on lib/basic/msg/js/en.js:
```diff
@@ -2,429 +2,433 @@
-/** @export */ Blockly.Msg.TEXTS_HUE = "160";
-/** @export */ Blockly.Msg.PROCEDURES_HUE = "290";
-/** @export */ Blockly.Msg.COLOUR_HUE = "20";
-/** @export */ Blockly.Msg.VARIABLES_DYNAMIC_HUE = "310";
```
Is it normal that those code lines are in the commit history?

jsccjj replied:
I think so since this is caused be using the latest "en.js" from the official Blockly repo. It seems Google changed the structure of "en.js". 

Original "en.js" in 1.1.0
![image](https://user-images.githubusercontent.com/79299554/119870022-c2c60180-bed5-11eb-92b9-fb42dc92534d.png)

New "en.js" from Google:
![image](https://user-images.githubusercontent.com/79299554/119870172-e8530b00-bed5-11eb-8369-ab3e95b11eea.png)

bartbutenaers replied:
Same comment as above: I was confusing my en.js with the one from Blockly.  So I thought: where the hell are those `@export` statements coming from ...  But now it makes sense ;-)

bartbutenaers reviewed (commented):
Hi Jeff (@jsccjj),

Sorry for the long delay!
Your code changes are very nice!  
I have added some remarks.  Once those are solved, I will merge the pull request.
Then I will add some other stuff, to make sure we have a complete release.
Thanks !!
Bart

jsccjj reviewed (commented):
> Hi Jeff (@jsccjj),
> 
> Sorry for the long delay!
> Your code changes are very nice!
> I have added some remarks. Once those are solved, I will merge the pull request.
> Then I will add some other stuff, to make sure we have a complete release.
> Thanks !!
> Bart

Hi Bart, 

No problem. I hope all is well over there!
I replied your comments. If you have any questions, please don't hesitate to let me know.

Thanks,

Jeff

bartbutenaers commented:
Jeff,

I have a series of new features for this Blockly node in an old local version of mine at home.  Would like to have those features also in the 1.1.0 release, so we can remove both old branches and my local changes.  Would like to have a clean Github repository again, with only a master branch.  That allows us to create smaller releases in the future, to avoid that we end up again with years of inactivity ...

Some of my old changes might not work anymore after your pull-request, so I will have to review those.  Which might take some time.

I will create a series of issues in this repository, so you can do a follow-up and get an idea of the progress...

Thanks!!
Bart
